apps/ejabberd/src/ejabberd_c2s.erl

```diff
@@ -2412,6 +2414,38 @@ flush_messages(N, Acc) ->
               {N, Acc}
     end.
 
+
+%%%----------------------------------------------------------------------
+%%% XEP-0016
+%%%----------------------------------------------------------------------
+
+maybe_update_presence(StateData = #state{jid = JID}, NewList) ->
+    {Froms, _Tos, _Pending} = ejabberd_hooks:run_fold(
```

Comment:
Moreover running the `roster_get_subscription_lists` hook will touch DB every time which is not the most efficient way :)